### ISSUE TITLE
Feature request : simplifying intents in forall loops (and friends)

### ISSUE THREAD

**Issue by npadmana** [chapel-lang/chapel#5512]
I often find myself doing multiple reductions of the same type in forall
loops (eg. multiple accumulators). The way I currently do this is to
specify separate reduce intents (as in the commented line in the code
below....)
```
  (+ reduce sum1, + reduce sum2)
```
While this works fine, it would be nice to write this as
```
  (+ reduce sum1,sum2)
```
or something like that (with a different delimiter, for instance). This could be extended to all intents (suggested by @nspark), instead of just reduce intents. 

@bradcray suggested using tuples for this 
```
  (in (x,y), + reduce (sum1, sum2))
```

I'll note that OpenMP allows specifying such common reductions over a list of variables.

This was discussed in chapel-users [here.](https://sourceforge.net/p/chapel/mailman/chapel-users/thread/alpine.LNX.2.00.1703040850190.1380%40sealcs01.us.cray.com/#msg35701864)

A possible future ---
```
var sum1, sum2 = 0;

config const N=100000;

forall ii in 1..N with (+ reduce (sum1,sum2)) {
  sum1 += ii;
  sum2 += ii*ii;
}

writeln(sum1);
writeln(sum2);
```

**Comment by bradcray** [comment]
Just to capture my reservations about the original proposal (comma-separation of identifiers implies shared intent) on this issue, let me bring over my comments from email:

> While I am sympathetic to the desire to reduce typing, I think a potential 
> drawback of this approach is that it would make task/forall intents more 
> different than function intents.  Specifically, in the function context:
> 
> ````chapel
> proc foo(in x, y) { ... }
> ````
> 
> says that 'x' is passed in by 'copy in' intent while 'y' is passed by the 
> default itent for its type.  It would seem odd to me to have distinct 
> interpretations in these two contexts.
> 
> A few other notes:
> 
> * there isn't a reduce intent for the function call context, so arguably
>    it would be OK to do something different for it in the task/forall
>    context.
> 
> * I can't recall whether there is a default/blank intent for the
>    task/forall context currenetly.  If there is, it doesn't add any value
>   per se (since anything not listed in the 'with' clause will be passed by
>    default intent since "passing" is via scoping in this context.  All the
>    same, it might be weird to have the lack of an intent mean one thing in
>    the function context and a completely different thing in the forall/task
>    context)

**Comment by bradcray** [comment]
Also, tagging @vasslitvinov on this due to his heavy involvement in task intents.

**Comment by e-kayrakli** [comment]
I wanted this feature today. Analogy between task intents and formal intents somewhat resonates with me. But I find the parallels between individual arguments/task-private variables (e.g. `ref` means the same thing in both contexts) to be more important. In terms of how to specify multiple task intents, I view this to be more akin to variable declarations because in my mind I am just defining my task-private variables that happens to have some intent qualifiers. I would argue that my standpoint is consistent with where we are today already, because you can do

```chpl
forall i in 1..10 with (var x: int)
```

to declare a task-private variable. But you can't

```chpl
proc foo(var x: int)
```

to declare a function-private variable. IOW, task intents already look more similar to variable declarations rather than argument "declarations" in some respects.

From that standpoint, I prefer having multi-task-private-variable definitions in the language syntax.

For context, my case looks like this:

```chpl
  var Arr1, Arr2, Arr3: [1..n] real;

  coforall loc in Locales with (+ reduce Arr1, + reduce Arr2, + reduce Arr3) do on loc {
    coforall gpu in here.gpus with (+ reduce Arr1, + reduce Arr2, + reduce Arr3) do on gpu {
```

I did bump into this in the past with non-reduce intents. Related to that experience, Brad says

> there isn't a reduce intent for the function call context, so arguably
it would be OK to do something different for it in the task/forall
context.

I don't agree. While I want to be able to define multiple task intents with comma-separated lists, I don't like the inconsistency implied by this approach.

**Comment by bradcray** [comment]
@e-kayrakli : I think you're saying you want to write your code as:

```chapel
  var Arr1, Arr2, Arr3: [1..n] real;

  coforall loc in Locales with (+ reduce Arr1, Arr2, Arr3) do on loc {
    coforall gpu in here.gpus with (+ reduce Arr1, Arr2, Arr3) do on gpu {
```

is that right?

Also, when you say:

> I don't agree. While I want to be able to define multiple task intents with comma-separated lists, I don't like the inconsistency implied by this approach.

I think you're saying you're OK with supporting comma-separated reduce intents, you wouldn't limit it to that case, but extend it to any task intent, is that right?

[my reason for checking is that most of your comment seemed to support Nikhil's proposal, but the last part made me second-guess that since it sounded like you didn't want to have comma-separated reduce intents]

**Comment by e-kayrakli** [comment]
> is that right?

Yes. The symmetry between the arrays' declarations and the intents are really appealing. I am also a supporter of 80-column coding :)

> I think you're saying you're OK with supporting comma-separated reduce intents, you wouldn't limit it to that case, but extend it to any task intent, is that right?

Yes.

In my own (hopefully more clear) words, supporting this _only_ for `reduce` intents but not others is a step in the wrong direction, in my opinion.
